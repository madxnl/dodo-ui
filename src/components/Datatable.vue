<template>
  <div :class="[$style.Datatable, contentLoading && $style.loading]">
    <table>
      <tbody>
        <tr :class="stickyHeader && $style.stickyHeader">
          <th v-if="showSelect" style="width: 0" scope="col">
            <Checkbox
              :indeterminate="visibleSelected.length > 0 && visibleSelected.length < rows.length"
              :model-value="visibleSelected.length > 0 && visibleSelected.length === rows.length"
              @update:model-value="toggleSelectAll"
            />
          </th>

          <th
            v-for="col in enabledColumns"
            :key="col.name"
            scope="col"
            :class="[canSortCol(col) ? $style.sortable : '', isSortCol(col) && $style.sortActive]"
            :style="[alignStyle(col), widthStyle(col)]"
            @click="toggleColumnSort(col)"
          >
            <span :class="$style.colName">
              <span>
                <!--
                @slot Column header
                @binding {name} string
                @binding {object} column
              -->
                <slot :name="`${slotName(col)}-header`" :column="col">{{ col.name }}</slot>
              </span>
              <Icon v-if="getColumnSortIcon(col)" :name="getColumnSortIcon(col)!" :class="$style.sortIcon" />
            </span>
          </th>
        </tr>

        <tr
          v-for="(row, i) in sortedItems"
          :key="i"
          :class="[rowClick && $style.clickableRow, isSelected(row) && $style.rowSelected]"
          @click="rowClick && rowClick(row)"
        >
          <td v-if="showSelect">
            <Checkbox :model-value="isSelected(row)" @update:model-value="toggleSelect(row)" />
          </td>
          <td v-for="col in enabledColumns" :key="col.name" :class="[]" :style="alignStyle(col)">
            <slot :name="slotName(col)" :row="row" :column="col">
              {{ getValue(col, row) }}
            </slot>
          </td>
        </tr>

        <tr v-if="!rows.length" :class="$style.noResults">
          <td colspan="999">No results</td>
        </tr>

        <tr v-if="rows.length && showMore" :class="$style.showMore">
          <td colspan="999">
            <Button variant="text" color="info" @click="showMore">Show more results</Button>
          </td>
        </tr>

        <tr v-if="showFooter" :class="[$style.footer, stickyHeader && $style.stickyFooter]">
          <td v-if="showSelect" />
          <th v-for="col in enabledColumns" :key="col.name" :style="alignStyle(col)" scope="col">
            <slot :name="`${slotName(col)}-footer`" :column="col" />
          </th>
        </tr>
      </tbody>
    </table>
  </div>
</template>
<script lang="ts" setup generic="T">
import { computed, ref, watch, watchEffect } from 'vue'
import { Button, Checkbox, Icon } from '.'

export interface DatatableColumn<T = object> {
  name: string
  value?: keyof T | ((row: T) => unknown)
  align?: 'start' | 'end'
  width?: string
  sort?: string
  disabled?: boolean
  slot?: string
}

defineSlots<{
  [key: string]: (scope: { row: T; column: DatatableColumn<T> }) => any
  [key: `${string}-header`]: (scope: { column: DatatableColumn<T> }) => any
  [key: `${string}-footer`]: (scope: { column: DatatableColumn<T> }) => any
}>()

const props = defineProps<{
  rows: T[]
  columns: DatatableColumn<T>[]
  selection?: T[]
  rowClick?: (row: T) => void
  selectBy?: keyof T
  stickyHeader?: boolean
  contentLoading?: boolean
  showFooter?: boolean
  showMore?: () => Promise<unknown>
  sortValue?: string
  sortAsync?: (sortBy: string | undefined) => Promise<unknown>
}>()

const emit = defineEmits<{
  (e: 'update:selection', selection: unknown[]): void
  (e: 'update:sortValue', order: string | undefined): void
}>()

const sort = ref<string>()
const selected = ref<unknown[]>([])
const sortingAsync = ref(false)

watchEffect(() => {
  sort.value = props.sortValue
})
watchEffect(() => {
  selected.value = props.selection ?? []
})
watch(sort, (v) => {
  emit('update:sortValue', v)
})
watch(selected, (v) => {
  emit('update:selection', v as T[])
})

const sortReverse = computed(() => sort.value?.startsWith('-'))
const enabledColumns = computed(() => props.columns.filter((c) => !c.disabled))
const showSelect = computed(() => !!props.selection)

const sortedItems = computed(() => {
  const sortCol = enabledColumns.value.find(isSortCol)
  if (props.sortAsync || !sortCol) return props.rows
  return props.rows.slice().sort((a: any, b: any) => {
    const x = getValue(sortCol, a)
    const y = getValue(sortCol, b)
    const numbers = typeof x === 'number' && typeof y === 'number'
    const d = numbers ? y - x : `${y}`.localeCompare(`${x}`)
    return sortReverse.value ? d : -d
  })
})

function getColumnSortIcon(col: DatatableColumn<T>) {
  if (!isSortCol(col)) return undefined
  return sortReverse.value ? 'arrow_drop_up' : 'arrow_drop_down'
}

function isSortCol(col: DatatableColumn<T>) {
  return sort.value?.replace('-', '') === col.sort
}

function slotName(col: DatatableColumn<T>) {
  return col.slot || col.name.trim().replace(/\W/g, '')
}

function canSortCol(col: DatatableColumn<T>) {
  return !!col.sort
}

function alignStyle(col: DatatableColumn<T>) {
  return col.align ? `text-align:${col.align}` : ''
}

function widthStyle(col: DatatableColumn<T>) {
  return col.width ? `width:${col.width}` : ''
}

async function toggleColumnSort(col: DatatableColumn<T>) {
  if (!canSortCol(col)) return
  if (isSortCol(col)) {
    sort.value = sortReverse.value ? undefined : '-' + col.sort
  } else {
    sort.value = col.sort
  }
  if (props.sortAsync) {
    sortingAsync.value = true
    await props.sortAsync(sort.value).finally(() => {
      sortingAsync.value = false
    })
  }
}

function getRowSelectValue(row: any) {
  return props.selectBy ? row[props.selectBy] : row
}

function toggleSelect(row: unknown) {
  const id = getRowSelectValue(row)
  if (selected.value.includes(id)) {
    selected.value = selected.value.filter((x) => x !== id)
  } else {
    selected.value = selected.value.concat(id)
  }
}

function getValue(col: DatatableColumn<T>, row: any) {
  if (typeof col.value === 'function') return col.value(row)
  if (typeof col.value === 'string') return row[col.value]
  return ''
}

function isSelected(row: unknown) {
  if (!showSelect.value) return false
  const id = getRowSelectValue(row)
  return selected.value.includes(id)
}

const visibleSelected = computed(() => {
  if (!showSelect.value) return []
  return props.rows.filter((r) => selected.value.includes(getRowSelectValue(r)))
})

function toggleSelectAll() {
  const allSelected = visibleSelected.value.length === props.rows.length
  selected.value = allSelected ? [] : props.rows.map(getRowSelectValue)
}
</script>

<style module>
.Datatable {
  font: var(--dodo-font-base);
  color: var(--dodo-color-foreground);
  background: var(--dodo-color-background);
  display: flex;
  flex-flow: column;
  gap: var(--dodo-gap-4);
}
.Datatable table {
  border-collapse: collapse;
  border-spacing: 0;
  width: 100%;
}
.Datatable th {
  text-align: start;
  font-weight: var(--dodo-weight-bold);
}
.colName {
  display: flex;
  gap: 4px;
  align-items: center;
}
.colName > :first-child {
  flex-grow: 1;
}
.sortable {
  cursor: pointer;
  user-select: none;
  transition: all 0.1s;
}
.sortIcon {
  transition: all 0.1s;
  opacity: 0;
}
.sortable:hover {
  background: rgba(0, 0, 0, 0.03);
}
.sortable.sortActive .sortIcon {
  opacity: 1;
}
.Datatable th,
.Datatable td {
  padding: 8px 12px;
  box-sizing: border-box;
}
.noResults,
.showMore {
  text-align: center;
  height: 100px;
}
.showMore button {
  width: 100%;
}
.Datatable tr:not(:last-child) {
  border-bottom: 1px solid rgba(0, 0, 0, 0.1);
}
.clickableRow:hover {
  cursor: pointer;
  box-shadow: 0 1px 3px rgba(0, 0, 0, 0.2);
}
.Datatable .rowSelected {
  background: color-mix(in hsl, var(--dodo-color-info) 30%, transparent);
}
.stickyHeader,
.stickyFooter {
  position: sticky;
  box-shadow: 0 1px 3px rgba(0, 0, 0, 0.2);
  background: var(--dodo-color-background);
  z-index: 1;
}
.stickyHeader {
  top: 0;
}
.stickyFooter {
  bottom: 0;
}
.loading {
  pointer-events: none;
  opacity: 0.5;
}
</style>
