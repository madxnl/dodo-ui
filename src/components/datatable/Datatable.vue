<template>
  <table :class="$style.Datatable">
    <tr>
      <th
        v-for="col in enabledColumns" :key="col.name"
        :class="[
          $style.cell,
          col.value ? $style.sortable : '',
          isSortCol(col) && $style.sortActive,
        ]"
        :style="`text-align:${col.align};width:${col.width}`"
        @click="toggleColumnSort(col)"
      >
        <!-- <FormCheckbox
                v-if="selection && col==columns[0]"
                :model-value="everythingSelected" :indeterminate="!everythingSelected && someSelected"
                @update:model-value="toggleSelectAll" /> -->

        <span :class="$style.colName">
          <Text strong>{{ col.name }}</Text>
          <Icon v-if="col.value" :name="getColumnSortIcon(col)" :class="$style.sortIcon" />
        </span>
      </th>
    </tr>

    <tr
      v-for="(row, i) in sortedItems" :key="i" :class="[
        rowClick && $style.clickableRow,
      ]"
      @click="rowClick && rowClick(row)"
    >
      <td
        v-for="col in enabledColumns" :key="col.name" :class="[
          $style.cell,
          i + 1 === sortedItems.length && $style.lastRow,
        ]"
        :style="`text-align:${col.align}`"
      >
        <!-- <FormCheckbox
              v-if="selection && col==columns[0]"
              :model-value="itemSelected(row)" @update:model-value="toggleSelectItem(row)" /> -->
        <slot :name="col.slot ?? 'cell'" :row="row" :index="i" :column="col">
          {{ col.value ? col.value(row) : '' }}
        </slot>
      </td>
    </tr>
    <!-- <div v-if="$slots.tfoot">
      <span v-for="col,i in columns" :key="col.name" :class="colClass(i)">
         <FormCheckbox
              v-if="selection && col==columns[0]"
              :model-value="everythingSelected" :indeterminate="!everythingSelected && someSelected"
              @update:model-value="toggleSelectAll" />
        <slot name="tfoot" :column="col" :column-index="i" />
      </span>
    </div> -->
    <!-- <BaseDiv :padding="64">
      <BaseDetectVisible v-if="loadMore" v-model:visible="bottomVisible">
        <BaseSpinner />
      </BaseDetectVisible>
    </BaseDiv> -->

    <tr v-if="!rows.length" :class="$style.noResults">
      <td :colspan="columns.length">
        No results
      </td>
    </tr>
  </table>
</template>
<script lang="ts" setup>
import { computed, ref, VNode, watchEffect } from 'vue'
import { Icon, Text } from '..'

export interface Column<T> {
  name: string
  value?: (row: T) => string|number
  slot?: string
  align?: 'start'|'end'
  width?: string
}

// export interface DatatableSlots<T> {
//   cell?: (context: { row: T; column: Column<T> }) => Array<VNode> | undefined
// }

export type DatatableSlots<T> = Record<string, (context: { row: T; column: Column<T> }) => Array<VNode> | undefined>

export interface DatatableProps<T> {
  rows: T[]
  columns: (Column<T>|null)[]
  order?: string
  // selection?: unknown[]
  rowClick?: (row: T) => void
  // selectionValue?: (x: unknown) => unknown
  // loadMore?: () => void
  // updating?: boolean
}

const props = defineProps<DatatableProps<unknown>>()

const emit = defineEmits<{
  // (e: 'update:selection', selection: unknown[]): void
  (e: 'update:order', order: string|undefined): void
}>()

const order = ref<string>()
const orderReverse = ref(false)

watchEffect(() => { order.value = props.order })
watchEffect(() => { emit('update:order', order.value) })

const enabledColumns = computed(() => props.columns.filter(Boolean).map(c => c!))

const sortedItems = computed(() => {
  const sortCol = enabledColumns.value.find(isSortCol)
  if (!sortCol?.value) return props.rows
  return props.rows.slice().sort((a, b) => {
    const x = sortCol.value!(a)
    const y = sortCol.value!(b)
    const numbers = typeof x === 'number' && typeof y === 'number'
    const d = numbers ? x - y : `${y}`.localeCompare(`${x}`)
    return orderReverse.value ? d : -d
  })
})

function getColumnSortIcon(col: Column<unknown>) {
  return (isSortCol(col) && orderReverse.value) ? 'arrow_drop_up' : 'arrow_drop_down'
}

function isSortCol(col: Column<unknown>) {
  return order.value === col.name
}

function toggleColumnSort(col: Column<unknown>) {
  if (!col.value) return
  if (isSortCol(col)) {
    if (orderReverse.value) order.value = undefined
    orderReverse.value = !orderReverse.value
  } else {
    order.value = col.name
    orderReverse.value = false
  }
}

// function colHeaderClass(col: number) {
//   const { orderings } = props.columns[col]
//   return {
//     ...colClass(col),
//     '-sortable': orderings && orderings.length > 0,
//   }
// }

// function colClass(col: number) {
//   const { align } = props.columns[col]
//   return {
//     '-right': align === 'right',
//   }
// }

// function itemSelectValue(row: unknown) {
//   return props.selectionValue ? props.selectionValue(row) : row
// }

// function itemSelected(row: unknown) {
//   return !!props.selection?.includes(itemSelectValue(row))
// }

// function toggleSelectItem(row: unknown) {
//   const value = itemSelectValue(row)
//   if (props.selection?.includes(value)) {
//     emit('update:selection', props.selection.filter(x => x !== value))
//   } else {
//     emit('update:selection', [...props.selection || [], value])
//   }
// }

// const everythingSelected = computed(() => {
//   return props.items.every(itemSelected) && props.items.some(itemSelected)
// })

// const someSelected = computed(() => {
//   return props.items.some(itemSelected)
// })

// function toggleSelectAll(toggle: boolean|undefined) {
//   emit('update:selection', toggle ? props.items.map(itemSelectValue) : [])
// }

// let loadMoreDelayTimer: ReturnType<typeof setTimeout>
// watchEffect(() => {
//   clearTimeout(loadMoreDelayTimer)
//   if (bottomVisible.value && props.loadMore) {
//     loadMoreDelayTimer = setTimeout(() => {
//       if (props.loadMore) props.loadMore()
//     }, 300)
//   }
// })
</script>

<style module>
.Datatable {
  font: var(--dodo-font-base);
  background: var(--dodo-color-background);
  /* color: black; */
  border-collapse: collapse;
  border-spacing: 0;
  width: 100%;
}
.Datatable th {
  text-align: start;
}
.colName {
  display: flex;
  gap: 8px;
  align-items: center;
}
.colName > :first-child {
  flex-grow: 1;
}
.sortable {
  cursor: pointer;
  user-select: none;
  transition: all .1s;
}
.sortIcon {
  transition: all .1s;
  opacity: 0;
}
.sortable:hover {
  background: rgba(0,0,0,0.03);
}
/* .sortable:hover .sortIcon {
  opacity: 0.5;
} */
.sortable.sortActive .sortIcon {
  opacity: 1;
}
.cell {
  padding: 8px 12px;
  /* display: flex; */
  align-items: center;
  gap: 8px;
}
.cell {
  border-bottom: 1px solid rgba(0,0,0,0.15);
}
.noResults {
  grid-row: span var(--n-cols);
  padding: 8px;
}
.lastRow {
  border: none;
}
.clickableRow:hover {
  cursor: pointer;
  background: rgba(0,0,0,0.03);
}
/*
  td, th {
    padding: 8px 16px;
    text-align: left;
    max-width: 400px;
    background: white;
  }
  tbody {
    td, th {
      height: 64px;
      border-bottom: 1px solid var(--color-outline-grey-medium);
    }
  }
  thead, tfoot {
    td, th {
      background: var(--color-grey-100);
      height: 48px;
      color: var(--color-text-black-medium);
      font: var(--font-body-medium);
      text-transform: capitalize;
      border-top: 1px solid var(--color-outline-grey-medium);
      border-bottom: 1px solid var(--color-outline-grey-medium);
    }
  }
  tr > :first-child {
    left: 0;
    position: sticky;
  }
  thead {
    top: 0;
    position: sticky;
    z-index: 1;
  }
  tfoot {
    bottom: 0;
    position: sticky;
    z-index: 1;
  }

  &.-clickable tbody tr,
  th.-sortable {
    cursor: pointer;
    &:hover td { background: var(--color-grey-100); }
    &:active td { background: var(--color-grey-200); }
  }

  tr.-selected td { background: rgba(var(--rgb-info-500), 0.2) !important; }
}
.InterfaceDatatable_cell {
  display: grid;
  grid-gap: 8px;
  grid-auto-flow: column;
  align-items: center;
  justify-content: start;
}
td, th {
  &.-right .InterfaceDatatable_cell {
    justify-content: end;
  }
} */
</style>
