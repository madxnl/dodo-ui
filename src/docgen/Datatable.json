{
  "exportName": "default",
  "displayName": "Datatable",
  "description": "",
  "tags": {},
  "props": [
    {
      "name": "rows",
      "required": true,
      "type": {
        "name": "Array",
        "elements": [
          {
            "name": "T"
          }
        ]
      }
    },
    {
      "name": "columns",
      "required": true,
      "type": {
        "name": "Array",
        "elements": [
          {
            "name": "DatatableColumn",
            "elements": [
              {
                "name": "T"
              }
            ]
          }
        ]
      }
    },
    {
      "name": "selection",
      "required": false,
      "type": {
        "name": "Array",
        "elements": [
          {
            "name": "T"
          }
        ]
      }
    },
    {
      "name": "rowClick",
      "required": false,
      "type": {
        "name": "TSFunctionType"
      }
    },
    {
      "name": "selectBy",
      "required": false,
      "type": {
        "name": "TSTypeOperator"
      }
    },
    {
      "name": "stickyHeader",
      "required": false,
      "type": {
        "name": "boolean"
      }
    },
    {
      "name": "contentLoading",
      "required": false,
      "type": {
        "name": "boolean"
      }
    },
    {
      "name": "showFooter",
      "required": false,
      "type": {
        "name": "boolean"
      }
    },
    {
      "name": "showMore",
      "required": false,
      "type": {
        "name": "TSFunctionType"
      }
    },
    {
      "name": "sortValue",
      "required": false,
      "type": {
        "name": "string"
      }
    },
    {
      "name": "sortAsync",
      "required": false,
      "type": {
        "name": "TSFunctionType"
      }
    }
  ],
  "events": [
    {
      "name": "update:selection",
      "type": {
        "names": [
          "Array"
        ],
        "elements": [
          {
            "name": "unknown"
          }
        ]
      }
    },
    {
      "name": "update:sortValue",
      "type": {
        "names": [
          "union"
        ],
        "elements": [
          {
            "name": "string"
          },
          {
            "name": "undefined"
          }
        ]
      }
    }
  ],
  "slots": [
    {
      "name": "`${slotName(col)}-header`",
      "scoped": true,
      "description": "Column header",
      "bindings": [
        {
          "name": "name",
          "title": "binding"
        },
        {
          "title": "binding",
          "type": {
            "name": "object"
          },
          "name": "column"
        }
      ]
    },
    {
      "name": "slotName(col)",
      "scoped": true,
      "bindings": [
        {
          "name": "name",
          "title": "binding"
        },
        {
          "name": "row",
          "title": "binding"
        },
        {
          "name": "column",
          "title": "binding"
        }
      ]
    },
    {
      "name": "`${slotName(col)}-footer`",
      "scoped": true,
      "bindings": [
        {
          "name": "name",
          "title": "binding"
        },
        {
          "name": "column",
          "title": "binding"
        }
      ]
    }
  ],
  "sourceFiles": [
    "src/components/Datatable.vue"
  ]
}