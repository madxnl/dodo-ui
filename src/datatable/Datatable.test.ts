import { mount, VueWrapper } from '@vue/test-utils'
import { expect, test, vi } from 'vitest'
import Datatable from './DatatableTyped.vue'

const rows = [
  { id: 1, name: 'Foo', modified: '2022-01-01' },
  { id: 2, name: 'Bar', modified: '2022-02-01' },
  { id: 3, name: 'Baz', modified: '2022-01-04' },
]

const columns = [
  { name: 'Name', value: 'name', sort: 'name' },
  { name: 'Modified', value: 'modified' },
]

const findByText = (w: VueWrapper, s: string) => w.findAll('*').find(w => w.text() === s)!

test('Column sort', async () => {
  const wrapper = mount(Datatable, { props: { columns, rows } })
  await findByText(wrapper, 'Name').trigger('click')
  expect(wrapper.emitted('update:sortValue')![0]).toEqual(['name'])
  await findByText(wrapper, 'Name').trigger('click')
  expect(wrapper.emitted('update:sortValue')![1]).toEqual(['-name'])
  await findByText(wrapper, 'Name').trigger('click')
  expect(wrapper.emitted('update:sortValue')![2]).toEqual([undefined])
  expect(wrapper.html()).toMatchSnapshot()
})

test('Selecting rows', async () => {
  const wrapper = mount(Datatable, { props: { columns, rows, selection: [], selectBy: 'id' } })
  const checkboxes = wrapper.findAllComponents({ name: 'Checkbox' })
  await checkboxes[0].trigger('click')
  expect(wrapper.emitted('update:selection')![0]).toEqual([[1, 2, 3]])
  await checkboxes[1].trigger('click')
  expect(wrapper.emitted('update:selection')![1]).toEqual([[2, 3]])
  await checkboxes[1].trigger('click')
  expect(wrapper.emitted('update:selection')![2]).toEqual([[2, 3, 1]])
  expect(wrapper.html()).toMatchSnapshot()
})

test('Show more', async () => {
  const showMore = vi.fn(async () => {})
  const wrapper = mount(Datatable, { props: { columns, rows, showMore } })
  await findByText(wrapper, 'Show more results').find('button').trigger('click')
  expect(showMore).toBeCalled()
  expect(wrapper.html()).toMatchSnapshot()
})
