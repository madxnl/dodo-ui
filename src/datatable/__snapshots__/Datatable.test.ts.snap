// Vitest Snapshot v1

exports[`Column sort 1`] = `
"<div class=\\"_Datatable_9c1ce1\\">
  <table>
    <tr class=\\"\\">
      <!--v-if-->
      <th scope=\\"col\\" class=\\"_sortable_9c1ce1\\"><span class=\\"_colName_9c1ce1\\"><span>Name</span>
        <!--v-if--></span>
      </th>
      <th scope=\\"col\\" class=\\"_sortActive_9c1ce1\\"><span class=\\"_colName_9c1ce1\\"><span>Modified</span><span class=\\"_Icon_665cc9 material-symbols-outlined _sortIcon_9c1ce1\\">arrow_drop_down</span></span></th>
    </tr>
    <tr class=\\"\\">
      <!--v-if-->
      <td class=\\"\\" style=\\"\\">Foo</td>
      <td class=\\"\\" style=\\"\\">2022-01-01</td>
    </tr>
    <tr class=\\"\\">
      <!--v-if-->
      <td class=\\"\\" style=\\"\\">Baz</td>
      <td class=\\"\\" style=\\"\\">2022-01-04</td>
    </tr>
    <tr class=\\"\\">
      <!--v-if-->
      <td class=\\"\\" style=\\"\\">Bar</td>
      <td class=\\"\\" style=\\"\\">2022-02-01</td>
    </tr>
    <!--v-if-->
    <!--v-if-->
    <!--v-if-->
  </table>
</div>"
`;

exports[`Selecting rows 1`] = `
"<div class=\\"_Datatable_9c1ce1\\">
  <table>
    <tr class=\\"\\">
      <th style=\\"width: 0px;\\" scope=\\"col\\"><span class=\\"_Checkbox_01f563 _checked_01f563\\"><span class=\\"_Icon_665cc9 material-symbols-outlined\\">check_box</span><input type=\\"checkbox\\"></span></th>
      <th scope=\\"col\\" class=\\"_sortable_9c1ce1\\"><span class=\\"_colName_9c1ce1\\"><span>Name</span>
        <!--v-if--></span>
      </th>
      <th scope=\\"col\\" class=\\"_sortActive_9c1ce1\\"><span class=\\"_colName_9c1ce1\\"><span>Modified</span><span class=\\"_Icon_665cc9 material-symbols-outlined _sortIcon_9c1ce1\\">arrow_drop_down</span></span></th>
    </tr>
    <tr class=\\"_rowSelected_9c1ce1\\">
      <td><span class=\\"_Checkbox_01f563 _checked_01f563\\"><span class=\\"_Icon_665cc9 material-symbols-outlined\\">check_box</span><input type=\\"checkbox\\"></span></td>
      <td class=\\"\\" style=\\"\\">Foo</td>
      <td class=\\"\\" style=\\"\\">2022-01-01</td>
    </tr>
    <tr class=\\"_rowSelected_9c1ce1\\">
      <td><span class=\\"_Checkbox_01f563 _checked_01f563\\"><span class=\\"_Icon_665cc9 material-symbols-outlined\\">check_box</span><input type=\\"checkbox\\"></span></td>
      <td class=\\"\\" style=\\"\\">Baz</td>
      <td class=\\"\\" style=\\"\\">2022-01-04</td>
    </tr>
    <tr class=\\"_rowSelected_9c1ce1\\">
      <td><span class=\\"_Checkbox_01f563 _checked_01f563\\"><span class=\\"_Icon_665cc9 material-symbols-outlined\\">check_box</span><input type=\\"checkbox\\"></span></td>
      <td class=\\"\\" style=\\"\\">Bar</td>
      <td class=\\"\\" style=\\"\\">2022-02-01</td>
    </tr>
    <!--v-if-->
    <!--v-if-->
    <!--v-if-->
  </table>
</div>"
`;

exports[`Show more 1`] = `
"<div class=\\"_Datatable_9c1ce1\\">
  <table>
    <tr class=\\"\\">
      <!--v-if-->
      <th scope=\\"col\\" class=\\"_sortable_9c1ce1\\"><span class=\\"_colName_9c1ce1\\"><span>Name</span>
        <!--v-if--></span>
      </th>
      <th scope=\\"col\\" class=\\"_sortActive_9c1ce1\\"><span class=\\"_colName_9c1ce1\\"><span>Modified</span><span class=\\"_Icon_665cc9 material-symbols-outlined _sortIcon_9c1ce1\\">arrow_drop_down</span></span></th>
    </tr>
    <tr class=\\"\\">
      <!--v-if-->
      <td class=\\"\\" style=\\"\\">Foo</td>
      <td class=\\"\\" style=\\"\\">2022-01-01</td>
    </tr>
    <tr class=\\"\\">
      <!--v-if-->
      <td class=\\"\\" style=\\"\\">Baz</td>
      <td class=\\"\\" style=\\"\\">2022-01-04</td>
    </tr>
    <tr class=\\"\\">
      <!--v-if-->
      <td class=\\"\\" style=\\"\\">Bar</td>
      <td class=\\"\\" style=\\"\\">2022-02-01</td>
    </tr>
    <!--v-if-->
    <tr class=\\"_showMore_9c1ce1\\">
      <td colspan=\\"999\\"><button style=\\"--bnt-rgb: var(--dodo-rgb-info);\\" type=\\"button\\" class=\\"_Button_962483 _text_962483 _loading_962483\\">
          <div class=\\"_content_962483\\">Show more results</div>
          <div class=\\"_Spinner_b556f8 _spinner_962483\\"><svg style=\\"\\" viewBox=\\"0 0 18 18\\">
              <path stroke=\\"currentColor\\" fill=\\"none\\" d=\\"M 9 1 A 8 8 0 0 1 17 9\\" stroke-width=\\"2\\"></path>
              <path stroke=\\"currentColor\\" opacity=\\"0.33\\" fill=\\"none\\" d=\\"M 17 9 A 8 8 0 1 1 9 1\\" stroke-width=\\"2\\"></path>
            </svg></div>
        </button></td>
    </tr>
    <!--v-if-->
  </table>
</div>"
`;
